Set toolchain clang linker to gold (Linux)

etcwilde opened the pull request:
BFD doesn't work with Swift symbols. We get nasty errors like this:

```
error: link command failed with exit code 1 (use -v to see invocation)
/usr/bin/ld: ...wift/CMakeFiles/swiftDispatch.dir/Block.swift.o:
  relocation R_X86_64_PC32 against protected symbol
  `$s8Dispatch0A13WorkItemFlagsVSYAAMc' can not be used when making a
  shared object
/usr/bin/ld: final link failed: bad value
```

rdar://123061492

etcwilde committed:
Set toolchain clang linker to gold

BFD doesn't work with Swift symbols. We get nasty errors like this:

```
error: link command failed with exit code 1 (use -v to see invocation)
/usr/bin/ld: ...wift/CMakeFiles/swiftDispatch.dir/Block.swift.o:
  relocation R_X86_64_PC32 against protected symbol
  `$s8Dispatch0A13WorkItemFlagsVSYAAMc' can not be used when making a
  shared object
/usr/bin/ld: final link failed: bad value
```

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test Linux toolchain

etcwilde commented:
https://github.com/apple/swift-driver/pull/1596
@swift please test Linux

etcwilde commented:
https://github.com/apple/swift-driver/pull/1596
@swift-ci please test Linux

compnerd reviewed (approved):
Yeah, that is a long standing issue with BFD and our usage of protected symbols.

al45tair commented:
https://github.com/apple/swift/pull/73474
@swift-ci Please test macOS platform

etcwilde committed:
Use consistent clang through build

The just-built Swift driver was picking up the system clang, which would
try to use the bfd linker instead of gold or lld. bfd doesn't like how
Swift handles protected symbols and fails with the following error:

```
error: link command failed with exit code 1 (use -v to see invocation)
/usr/bin/ld: /home/build-user/build/buildbot_linux/libdispatch-linux-x86_64/src/swift/CMakeFiles/swiftDispatch.dir/Block.swift.o: relocation R_X86_64_PC32 against protected symbol `$s8Dispatch0A13WorkItemFlagsVSYAAMc' can not be used when making a shared object
/usr/bin/ld: final link failed: bad value
```

This patch tells the Swift-Driver where it should find its clang, and
makes it match the clang that is in use through the rest of the build.

etcwilde commented:
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Linux platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Linux platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Linux platform

etcwilde commented:
IT'S FINALLY GREEN ON LINUX! 😆

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test macOS Platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Windows platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Linux platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Linux platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Linux platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Linux platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test Windows Platform

etcwilde commented:
https://github.com/apple/llvm-project/pull/8748
https://github.com/apple/swift-driver/pull/1608
@swift-ci please test macOS Platform
